torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp

```diff
@@ -117,25 +118,35 @@ static bool needsGradientInProfilingMode(Block* b) {
   return false;
 }
 
-void guardDifferentiableGraph(Node* dnode) {
+bool guardDifferentiableGraph(Node* dnode) {
   auto gi = dnode->g(attr::Subgraph)->inputs();
+  bool all_requires_grad_defined = true;
   for (size_t i = 0; i < gi.size(); i++) {
     auto ty = gi[i]->type()->cast<TensorType>();
     if (ty) {
       auto n = gi[i]->uses().at(0).user;
+      auto dni = dnode->inputs().at(i);
       if (n->kind() == prim::profile) {
-        dnode->inputs().at(i)->setType(n->ty(attr::profiled_type));
+        dni->setType(n->ty(attr::profiled_type));
       }
+      // we check if the optional is defined
+      all_requires_grad_defined &=
+          dni->type()->cast<TensorType>()->requiresGrad().has_value();
     }
   }
-  insertTypeGuard(dnode, [](const TensorTypePtr& t) {
-    TORCH_INTERNAL_ASSERT(
-        t->requiresGrad(),
-        "profiler did not say whether grad is required"); // i.e. we want the
-                                                          // optional to be
-                                                          // defined
-    return TensorType::get()->withRequiresGrad(t->requiresGrad());
-  });
+  if (all_requires_grad_defined) {
```

Comment:
it's somewhat common for LSTMs (in a loop) to have tensor values whose `requires_grad` attributes flip from `false` (in the first iteration) to `true` (in the subsequent iterations). This behaviour would result in `all_requires_grad_defined` set to `false`. The other probable scenario would be when we switch from training to eval/test (i.e. `true` to `false`).

I think we could consider the following options:
a) enable re-profiling if we aren't sure, so we will profile a steady state (would cost us an extra call in a loop, to **always** go through a fallback)
b) gamble and assume that requires_gradients will be `true`. This strategy works out very nicely for both LSTMs and eval/train scenarios. We would take the fallback in the first iterations since `requires_gradients` would be false, but then we will be running the optimized path (no fallbacks). Similarly, we would spend more time in training than eval, so incurring the fallback penalty in the test/eval runs is preferable. 
c) peel the first iteration, so profiler only sees `requires_grad`=true. We kinda tried it but very reluctantly.
d) do nothing and lose out on perf since we won't fuse in non-diff regions as long as even a single `requires_grad`=true anywhere in a graph.

Thoughts
@t-vi @eellison @jjsjann123  ?

Reply:
Oha, indeed! I would go for b).

Reply:
I also kinda like b). It seems to be a reasonable heuristic; we could always revise it later.

> This will apply to completely unseen things, too, though...

that's true. we could check for unseen things as their types are just `TensorType::get()`.

>  I'd probably go for actually querying the configuration in the ProfilingRecord for which is the most commonly seen.

I'm actually working on this in a separate PR. However, the initial version would aggregate across different invocations of a function rather than different iterations of a loop.

Reply:
> that's true. we could check for unseen things as their types are just TensorType::get().

I wonder whether we might just have a "seen" bit on the profiling node. I've been thinking about incremental profiling (where we leave profiling nodes for the bits we don't have records of and re-run the optimization once we hit them) and will probably play with it one of these days.

Reply:
Yea, I think we should do b) but filter out `TensorType::get()` (unseen tensors)

Reply:
OK, I'll return to the taking out graphs with TensorType::get() for now and look into incremental profiling. Thank you for the insightful comments!